core/src/km_core_state_context_set_if_needed.cpp

```diff
@@ -26,115 +26,226 @@
 
 using namespace km::core;
 
+enum context_changed_type {
+  // new and internal contexts are identical
+  CONTEXT_UNCHANGED = 0,
+  // new and internal contexts differ
+  CONTEXT_DIFFERENT = 1,
+  // internal context is shorter, but content is same as far as it goes
+  CONTEXT_SHORTER = 2,
+  // internal context is longer, but content is same as far as it goes
+  CONTEXT_LONGER = 3,
+};
+
+typedef struct {
+  uint8_t type;
+  // the number of characters/context items (from beginning of context)
+  // that are missing from the new/internal context.
+  uint8_t end_difference;
```

Comment:
Can we make `end_difference` a larger type? I don't think we need to limit ourselves here for the future.